Add instructions to deploy the webhook to a Google Cloud Function

SimonBarendse opened the pull request.

SimonBarendse committed:
Add instructions to deploy the webhook to a Google Cloud Function

jpcoenen commented on README.md:
```diff
@@ -9,3 +9,25 @@ It detects whether the container environment contains secret references (`secret
+We're also [working on](https://github.com/secrethub/secrethub-kubernetes-mutating-webhook/pull/2) a way to deploy the webhook in the Kubernetes cluster itself.
+
+You can deploy the webhook to a Google cloud function using the following steps:
+
```
Don't you have to `git clone` first? Or where do I have to execue this?

SimonBarendse replied:
Yes, you have to clone first. I thought this is so obvious that we can leave it out. But because you're asking this question I'll add this step to make it extra clear.

jpcoenen commented on README.md:
````diff
@@ -9,3 +9,25 @@ It detects whether the container environment contains secret references (`secret
+
+1. Deploy the webhook to a Google Cloud Function:
+```sh
+gcloud functions deploy secrethub-mutating-webhook --runtime go113 --trigger-http
````
What if you get the question if you want to allow unauthenticated requests?

SimonBarendse replied:
I've allowed unauthenticated requests. Access to the function doesn't give you access to any secret data or resources, it just enables you to change a pod specification you have access to in a predefined way.

If you do want restricted access, I imagine using IAM or #2

SimonBarendse replied:
I'll add a note.

SimonBarendse replied:
done

jpcoenen commented on README.md:
````diff
@@ -9,3 +9,25 @@ It detects whether the container environment contains secret references (`secret
+
+1. Deploy the webhook to a Google Cloud Function:
+```sh
+gcloud functions deploy secrethub-mutating-webhook --runtime go113 --trigger-http
````
This results in the following error for me:
```
ERROR: (gcloud.functions.deploy) OperationError: code=3, message=Build failed: # serverless_function_app
./main.go:24:43: cannot refer to unexported name webhook.secrethub
./main.go:24:43: undefined: webhook.secrethub
./main.go:24:66: undefined: mutating
./main.go:24:75: undefined: webhook
./main.go:26:44: cannot refer to unexported name webhook.secrethub
./main.go:26:44: undefined: webhook.secrethub
./main.go:26:67: undefined: mutating
./main.go:26:76: undefined: webhook; Error ID: 0eec69b1
```

SimonBarendse replied:
Ah, yes, the entrypoint is missing:

```suggestion
gcloud functions deploy secrethub-mutating-webhook --runtime go113 --entry-point F --trigger-http
```

SimonBarendse committed:
Specify entrypoint in cloud function deploy

SimonBarendse committed:
Add deploy step to clone and cd into this repo

florisvdg commented on deploy.yaml:
```diff
@@ -0,0 +1,18 @@
+apiVersion: admissionregistration.k8s.io/v1beta1
+kind: MutatingWebhookConfiguration
+metadata:
+  name: secrethub-mutating-webhook
+  labels:
+    app: secrethub-mutating-webhook
+    kind: mutator
+webhooks:
+- name: secrethub-mutating-webhook.default.svc.cluster.local
+  clientConfig:
+    url: "YOUR_CLOUD_FUNCTION_URL"
+    caBundle: ""
+  rules:
+  - operations: ["CREATE"]
+    apiGroups: [""]
+    apiVersions: ["v1"]
+    resources: ["pods"]
+  failurePolicy: Fail
```
Let's do `deploy/gcloud-function/config.yml`.

florisvdg replied:
We could also move the deploy instructions `README` there if we want (`deploy/gcloud-function/README.md`).

florisvdg replied:
And later we can add `deploy/k8s/config.yml` plus `deploy/k8s/webhook.yml` and also `deploy/aws-lambda/config.yml` possibly.

SimonBarendse replied:
Yes, sounds good :+1:

SimonBarendse committed:
Move gcloud function deploy to a subdirectory

SimonBarendse committed:
Set --allow-unauthenticated when deploying cloud function

This is acceptable because the function doesn't give access to any
resources or data. It only allows you to mutate provided data.
